### PR TITLE
chore: Script for managing libs dependencies and imports

### LINKED ISSUE #10608: Create script for dependencies management
He have a lot of inconsistency in dependencies listed in each library and between versions of them.

We can automate that process.
- we keep all dependencies in root package.json
- we specify automatically all peerDependencies in libraries based on imports and we get the version from root package.json
- schematics and migrations then are based on each lib package.json, so everything is listed
- updates of packages in root will be populated in all relevant places

### PR DESCRIPTION
**Pull request by marlass**
### Idea behind the script
Keeping manual list of peerDependencies and version in each library is very error-prone and it can easily cause bugs. Instead we can have single source of truth got versions (root package.json) and the single source of truth for dependencies are the imports present in library source files (ts and scss). We can change our workflow with this script in few ways.
1. We have only one directory that we need to `yarn`.
2. CI can catch regression (using new dependency in minor/patch versions!)
3. Every spartacus lib uses the same version of libs.
4. Updating packages in root package.json affects everything. In major releases we need to bump only in package.json and then run `yarn config:update --breaking-changes`. All our packages have exactly the same version of peerDependencies.
5. Keeping unused peerDependencies will be catched by CI
6. Using `rxjs/internal` imports will be catched by CI
7. Using absolute paths will be catched by the CI
8. Using Node.js imports outside of SSR will be catched by CI
9. All tsconfig files are automatically adjusted for new dependencies.
10. Each library schematics can now use lib package.json as a reliable way of setting installed deps and their versions (also for migrations).
11. We keep implementing tooling for packages in generic ways (eg. using the same setup for building schematics, running unit tests, building json translation files).

~~In this PR I intentionally created few errors to showcase all checks (will remove them after first review and before merge).~~ Done

Second part to this PR is https://github.com/SAP/spartacus/pull/10739 which fixes all the errors catched by the tools and suppress some errors before we implement better solutions (json generation and external schematics deps). It's supposed to be merged after this one will land in develop.

Once all the PR's all merged we can enable config check in github actions as a required check.

Also before 4.0 we have to look once again into schematics and update the way we specify the dependencies and their version for install to pull the info from package.json.

Closes #10608

### CODE REVIEW

File: tsconfig.compodoc.json [hunk 1]
```diff
@@ -21,7 +21,7 @@
     "noUnusedParameters": true,
     "paths": {
       "@spartacus/schematics": [
-        "projects/schematics/src/public_api"
+        "projects/schematics/src/public_ap"
```
Comment (hackergil) [inline tsconfig.compodoc.json]: typo I think?

File: tools/config/tsconfig-paths.ts [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+  Repository,
+  success,
+} from './index';
+
+function readTsConfigFile(path: string): any {
+  return parse(fs.readFileSync(path, 'utf-8'));
+}
+
+function setCompilerOptionsPaths(tsconfigPath: string, paths: Object) {
+  const tsConfigContent = parse(fs.readFileSync(tsconfigPath, 'utf-8'));
```
Comment (hackergil) [inline tools/config/tsconfig-paths.ts]: call readTsConfigFile() instead?

**Review by hackergil — CHANGES_REQUESTED** [review]
- When I ran the script, I noticed that there are checks for some APIs in node_modules, which should not be the case I think?
- Some comments on the PR
- Build is not passing

File: tools/config/manage-dependencies.ts [hunk 3]
```diff
@@ -0,0 +1,1163 @@
...
+                  errorsFound = true;
+                  error(
+                    file,
+                    [
+                      `Absolute import should not be used outside of spec files.\n    Referenced \`${chalk.bold(
+                        imp.importPath
+                      )}\``,
+                    ],
+                    [
+                      `You should use absolute paths only for testing modules.`,
```
Comment (znikola) [inline tools/config/manage-dependencies.ts]: Maybe we phrase it like this:
`You should use absolute paths only in test files`?

File: tools/config/manage-dependencies.ts [hunk 4]
```diff
@@ -0,0 +1,1163 @@
...
+        }
+      });
+    });
+    if (errors.length) {
+      error(PACKAGE_JSON, errors, [
+        `All dependencies that are directly referenced should be specified as \`${chalk.bold(
+          'dependencies'
+        )}\` or \`${chalk.bold('devDependencies')}\`.`,
+        `Install them with \`${chalk.bold(
+          'yarn add <dependency-name> (--dev)'
```
Comment (znikola) [inline tools/config/manage-dependencies.ts]: As `--dev` is an optional argument, let's use the common syntax for optional args: `[--dev]`

**Review by znikola — CHANGES_REQUESTED** [review]

**Review by znikola — DISMISSED** [review]

**Comment by hackergil** [comment]
@marlass Here's an example of what I was saying. We should not be scanning node_modules

```
Checking imports of Node.js APIs

--- projects/schematics/node_modules/@angular-devkit/core/node/host.d.ts -------
 ✖  Node.js API `fs` is referenced.

 i  Node.js APIs can only be used in SSR code or in schematics specs.
    You might have wanted to import it from some library instead.
--------------------------------------------------------------------------------


--- projects/schematics/node_modules/@angular-devkit/core/node/testing/index.d.ts
 ✖  Node.js API `fs` is referenced.

 i  Node.js APIs can only be used in SSR code or in schematics specs.
    You might have wanted to import it from some library instead.
---------------------------------------------------------------------------------


--- projects/schematics/node_modules/@nguniversal/express-engine/schematics/install/files/__serverFileName@stripTsExtension__.ts
 ✖  Node.js API `fs` is referenced.

 i  Node.js APIs can only be used in SSR code or in schematics specs.
    You might have wanted to import it from some library instead.
```

**Review by hackergil — CHANGES_REQUESTED** [review]
Added a comment about scanning node_modules

**Comment by marlass** [comment]
> @marlass Here's an example of what I was saying. We should not be scanning node_modules
> 
> ```
> Checking imports of Node.js APIs
> 
> --- projects/schematics/node_modules/@angular-devkit/core/node/host.d.ts -------
>  ✖  Node.js API `fs` is referenced.
> 
>  i  Node.js APIs can only be used in SSR code or in schematics specs.
>     You might have wanted to import it from some library instead.
> --------------------------------------------------------------------------------
> 
> 
> --- projects/schematics/node_modules/@angular-devkit/core/node/testing/index.d.ts
>  ✖  Node.js API `fs` is referenced.
> 
>  i  Node.js APIs can only be used in SSR code or in schematics specs.
>     You might have wanted to import it from some library instead.
> ---------------------------------------------------------------------------------
> 
> 
> --- projects/schematics/node_modules/@nguniversal/express-engine/schematics/install/files/__serverFileName@stripTsExtension__.ts
>  ✖  Node.js API `fs` is referenced.
> 
>  i  Node.js APIs can only be used in SSR code or in schematics specs.
>     You might have wanted to import it from some library instead.
> ```

We no longer have node_modules in libraries directories. Isn't is some leftover node_modules dir in your local repository?

**Review by hackergil — APPROVED** [review]

**Pull request merged**